src/instructions/port.rs

```diff
@@ -95,84 +95,72 @@ impl PortWrite for u32 {
     }
 }
 
-/// A read only I/O port.
-pub struct PortReadOnly<T> {
-    port: u16,
-    phantom: PhantomData<T>,
-}
+/// A marker trait for access types which allow reading port values.
+pub trait PortReadAccess {}
 
-/// A write only I/O port.
-pub struct PortWriteOnly<T> {
-    port: u16,
-    phantom: PhantomData<T>,
-}
+/// A marker trait for access types which allow writing port values.
+pub trait PortWriteAccess {}
 
-/// An I/O port.
-pub struct Port<T> {
-    port: u16,
-    phantom: PhantomData<T>,
+/// An access marker type indicating that a port is only allowed to read values.
+#[derive(Debug)]
+pub struct ReadOnlyAccess {
+    _priv: (),
 }
 
-impl<T> PortReadOnly<T> {
-    /// Creates a read only I/O port with the given port number.
-    #[inline]
-    pub const fn new(port: u16) -> PortReadOnly<T> {
-        PortReadOnly {
-            port,
-            phantom: PhantomData,
-        }
-    }
-}
+impl PortReadAccess for ReadOnlyAccess {}
 
-impl<T: PortRead> PortReadOnly<T> {
-    /// Reads from the port.
-    ///
-    /// ## Safety
-    ///
-    /// This function is unsafe because the I/O port could have side effects that violate memory
-    /// safety.
-    #[inline]
-    pub unsafe fn read(&mut self) -> T {
-        T::read_from_port(self.port)
-    }
+/// An access marker type indicating that a port is only allowed to write values.
+#[derive(Debug)]
+pub struct WriteOnlyAccess {
+    _priv: (),
 }
 
-impl<T> PortWriteOnly<T> {
-    /// Creates a write only I/O port with the given port number.
-    #[inline]
-    pub const fn new(port: u16) -> PortWriteOnly<T> {
-        PortWriteOnly {
-            port,
-            phantom: PhantomData,
-        }
-    }
+impl PortWriteAccess for WriteOnlyAccess {}
+
+/// An access marker type indicating that a port is allowed to read or write values.
+#[derive(Debug)]
+pub struct ReadWriteAccess {
+    _priv: (),
 }
 
-impl<T: PortWrite> PortWriteOnly<T> {
-    /// Writes to the port.
-    ///
-    /// ## Safety
-    ///
-    /// This function is unsafe because the I/O port could have side effects that violate memory
-    /// safety.
-    #[inline]
-    pub unsafe fn write(&mut self, value: T) {
-        T::write_to_port(self.port, value)
-    }
+impl PortReadAccess for ReadWriteAccess {}
+
+impl PortWriteAccess for ReadWriteAccess {}
+
+/// An I/O port.
+///
+/// The port reads or writes values of type `T` and has read/write access specified by `A`.
+///
+/// Use the provided marker types or aliases to get a port type with the access you need:
+/// * `Port<T, ReadWriteAccess>` -> `PortReadWrite<T>`
+/// * `Port<T, ReadOnlyAccess>` -> `PortReadOnly<T>`
+/// * `Port<T, WriteOnlyAccess>` -> `PortWriteOnly<T>`
+pub struct Port<T, A> {
+    port: u16,
+    phantom: PhantomData<(T, A)>,
 }
 
-impl<T> Port<T> {
+/// A read-write I/O port.
+pub type PortReadWrite<T> = Port<T, ReadWriteAccess>;
```

Comment:
I'm not a big fan of this breaking change. I would prefer `PortGeneric<T,A>` which is specialized as: `Port<T>`, etc...

Given that users will commonly use `Port` and rarely use `PortGeneric`, keeping the smaller name makes more sense to me (and it avoids a breaking change).